decode/lgt2md.1
." $Id: lgt2md.1,v 1.1 2008/09/18 17:32:51 yoksas Exp $
.TH LGTMD 1 "2008/09/18 17:32:51 $"
.SH NAME
lgt2md \- Unidata LDM decoder for NLDN and USPLN lightning data
.SH SYNOPSIS
.HP
.ft B
lgt2md
.nh
\%[-v]
\%[-x]
\%[-l\ \fIlogpath\fP]
\%[-[f|F]\ \fIfile\fP] nn schema <Keywords>
.hy
.ft
.SH DESCRIPTION
.LP
This program is an LDM-compatible decoder for NLDN and USPLN lightning data.
It reads data observations from standard input or from a file and decodes
that data into a McIDAS \fBMD\fP file.
.LP
This program can be invoked by an entry in an LDM configuration file similar
to the following (example IDD feedtype assumes LDM v6.7+):
.sp
.ta 4n +4n +4n
.nf
LIGHTNING	^(9[3-9])([0-3][0-9][0-9])([0-2][0-9])([0-5][0-9])([0-5][0-9])
	PIPE	-close
	lgt2md -vl logs/ldm-mcidas.log
	-d data/mcidas 70 NLDN DIALPROD=LD \\1\\2 \\3\\400 DEV=CCN
.fi
.sp
.ta 4n +4n +4n
.nf
LIGHTNING	^USPLN1EX-(....)_(..)_(..)_(..)_(..)_..\.uspln$
	PIPE	-close
	lgt2md -vl logs/ldm-mcidas.log
	-d data/mcidas 120 NLDN DIALPROD=LX \\1\\2 \\3\\400 DEV=CCN
.fi
.RE
.LP
Additional examples of LDM pqact.conf entries can be found in the file
ldm-mcidas-pqact.conf in the \fBetc/\fP subdirectory of the ldm-mcidas
installation tree.
.fi
.RE
.LP
Since this program reads from standard input, it can also be invoked as a
stand-alone program, outside the LDM environment.
.fi
.RE
.SH OPTIONS
.TP
.BI "-d " datadir
Change to directory \fIdatadir\fP before decoding.  As a consequence, the
output file will appear in that directory and the \fBROUTE.SYS\fP and
\fBSYSKEY.TAB\fP files will be looked for and, if found, used, and updated
in that directory.  Note that \fBROUTE.SYS\fP and \fBSYSKEY.TAB\fP must be
both readable and writable by the user running \fBlgt2md\fP.
.TP
.BI "-f " path
Read from file \fIpath\fP rather than from standard input.
.TP
.BI "-F " path
Read from file \fIpath\fP rather than from standard input.  At exit, delete
the file.
.TP
.BI "-l " logpath
Log file pathname.
The program uses Unidata's \fBulog\fP(3) package to write error and diagnostic
messages.
The default is to write to the \fBsyslogd\fP(8) daemon.  If this option is
specified, however, then messages will be written to the file
\fIlogpath\fP unless that pathname is "-", in which case all messages will be
written to standard error.
.TP
.B -v
Verbose logging.
Informative messages (level LOG_INFO) are logged.
By default, only messages of severity LOG_NOTICE and greater are
logged. See the Unidata LDM Site  Manager's  Guide  for more information on 
message logging.
.TP
.B -x
Debug logging.
Debugging messages (level LOG_DEBUG) are logged. See the Unidata LDM Site 
Manager's  Guide  for more information on message logging.
.SH POSITIONAL PARAMETERS
.TP
.B nn 
McIDAS \fBMD\fP file base number.
Data will be stored in \fBMD\fP files that follow the standard McIDAS naming 
rule.  The MD file number is the combination of the last digit of the Julian 
day [1-366] and the base \fBMD\fP number.  If the last digit of the Julian day
is zero, the \fBMD\fP file number will be the base number plus 10.  It is 
recommended to use 71 as the base MD number.
.TP
.B schema
McIDAS \fBMD\fP file schema name.
The schema file defines how the data will be stored and accessed in the MD 
file.  The currently supported schema, NLDN, is supplied in the binary file 
\fBDCNLDN\fP.  This schema must be added to the McIDAS schemas file, 
\fBSCHEMA\fP, in 
order for lgt2md to run.  For instructions on adding an \fBMD\fP file 
schema to \fBSCHEMA\fP, refer to \fBSCHE\fP section in the McIDAS-X Users
Guide.
.SH KEYWORDS
.TP
.BI DIALPROD= "product_code [day] [time]"
McIDAS product code identifier.
The \fBDIALPROD\fP keyword tells \fBlgt2md\fP to look in the McIDAS file 
routing table,
\fBROUTE.SYS\fP, to get the output \fBMD\fP file base number for the product
identified as \fIproduct_code\fP.  The optional 
\fIday\fP and \fItime\fP
parameters are useful for tracing decoding status in the log file.
If \fBROUTE.SYS\fP does not exist in the output directory (see \fB-d\fP flag
above) the \fBnn\fP base positional parameter described earlier will be used. In
order for \fBROUTE.SYS\fP to be used and updated, it must be both readable
and writable by the user running \fBlgt2md\fP.
.TP
.BI DEV= XYZ
McIDAS information reporting.
The \fBDEV\fP keyword tells \fBlgt2md\fP which informational messages 
to write to the
log file.  The form of the value for this keyword is a three capital letter
sequence each letter of which can be a \fBC\fP (print message) or \fBN\fP 
(suppress 
message). The order-dependent values refer to: standard messages, error 
messages, and debug messages, respectively.  The sequence \fBDEV=CCN\fP is 
recommended.
.SH NOTES
.LP
The \fBDIALPROD\fP
keyword entry must refer to a \fBROUTE.SYS\fP product of type \fBMD\fP.  

The McIDAS file routing table \fBROUTE.SYS\fP must reside in the same 
directory as the schema file, \fBSCHEMA\fP, and the output \fBMD\fP file
(see \fB-d\fP option above) and be readable and writable by the user running
\fBlgt2md\fP.  It is recommended to use the product code \fBLD\fP 
in updating \fBROUTE.SYS\fP (for more information see the section on 
\fBROUTE\fP in the McIDAS-X Users Guide):

.nf
ROUTE ADD LD MD 71 71 CC=3 SYS=2051 2052 2053 "NLDN Lightning Flashes
ROUTE REL LD
.fi

\fBSCHEMA\fP must reside in the same directory as the output \fBMD\fP file 
(see \fB-d\fP option above), and it must be readable by the user running 
\fBlgt2md\fP.  \fBSCHEMA\fP must also be a McIDAS-X file, not an McIDAS-OS2
one.

If the McIDAS system key table, \fBSYSKEY.TAB\fP, exists in the same directory
as the output \fBMD\fP file, table entries specified by the \fBSYS=\fP 
keyword (illustrated in the example above) will be updated
if it is both readable and writable by the user running \fBlgt2md\fP.
.RE
.SH "SEE ALSO"
.LP
.BR ~ldm-mcidas/etc/ldm-mcidas-pqact.conf,
.BR ldmd (1),
.BR ulog (3),
.BR syslogd (8).


RELEASE_NOTES
VERSION	COMMENTS
-------	-------------------------------------------------------------

2012    2012-01-18

        Updated version number to 2012 to reflect small updates (e.g.,
        new src/port/cfortran/cfortran.h, modified src/decode/Makefile)
        and that the package is still viable.

2008    2009-04-30

        Note: The version number is kept at 2008 since McIDAS went
        into maintenance mode with v2008.

        A variety of changes to C routines, Makefiles and test data
        files.

        Introduction of a new McIDAS lightning decoder, 'lgt2md', that
        handles both NLDN binary and USPLN ASCII data record formats.
        'lgt2md' is designed as a drop-in replacement for 'nldn2md':
        user's pattern action file entries for decoding NLDN data
        can be easily updated by replacing 'nldn2md' with 'lgt2md'.

        ROUTE.SYS (distributed as route.sys) was updated to include
        a new product, 'LP', for USPLN lightning data.

        The Makefile in the src/decode directory was reorganized to
        build the subset of the encoders/decoders contained in the package
        that are written entirely in C:

        area2png    - PNG-compresses the image portion of single-banded
                      McIDAS AREA files
        pnga2area   - uncompresses images created by 'area2png'
        zlibg2gini  - uncompresses the Zlib-compressed images contained
                      in the NOAAPort broadcast (and conveyed in the
                      IDD in the NIMAGE feed)
        pngg2gini   - uncompress NOAAPort images that have been
                      PNG-compressed by the Unidata NOAAPort ingest
                      package (archaic)
        proftomd    - decodes wind profiler netCDF files from NOAA/GSD
                      into McIDAS MD format files
        lgt2md      - decodes NLDN and USPLN lightning data into
                      McIDAS MD files

        This set is built by running 'make' or 'make basic'.
        
        In addition to this set, the Fortran-based 'nldn2md' NLDN
        lightning decoder will be built by 'make full'.

        The Makefile in src/decode also provides for the (attempted)
        building of archaic McIDAS decoders:

        cdftomd     - decodes wind profiler netCDF files from NOAA/GSD
                      into McIDAS MD format files (use 'proftomd' instead)
        gunrv2      - decodes compressed McIDAS GRID files sent in the
                      old Zephyr/Alden satellite broadcast into McIDAS
                      GRID format
        lwfile
        lwtmd2      - decodes compressed McIDAS MD files sent in the
                      old Zephyr/Alden satellite broadcast into McIDAS
                      MD format
        lwtoa3      - decodes delta-compressed McIDAS image files from
                      the old Zephyr/Alden satellite broadcast into McIDAS
                      AREA format
        nids2area   - converts Nexrad Level III images into McIDAS
                      AREA file format (best to use McIDAS IMGCOPY from
                      NEXR to AREA datasets directly as McIDAS capabilities
                      are kept up to date)
 
        In addition decoders built by 'make full', an attepmt will be
        made to build these decoders when the user runs 'make all'.
        There is _no_ guarantee that these decoders will be built
        successfully!

2004    2004-09-14

        Added O_TRUNC flag to mkdirs_open call in pnga2area.c,
        pngg2gini.c, and zlibg2gini.c.  This corrects a problem
        encountered when decoding Unidata-Wisconsin images (IDD
        feedtype UNIWISC aka MCIDAS) when using McIDAS routing table
        (ROUTE.SYS) concepts.  The problem was manifested in decoded
        image files keeping the size of the images they were
        overwriting.  So, the file size was always the "high water
        mark" of sizes for the products.

        NOTES:
        - this did not affect the sizes of files decoded without using
          McIDAS routing table concepts
        - the decoded images using McIDAS routing table concepts were
          valid/usable even though they are larger than they need to
          be

        Changed IDD datastream name from MCIDAS to UNIWISC in example
        pqact.conf files.

        Updated distribution with the versions of SATANNOT, SATBAND,
        and SCHEMA from the latest Unidata McIDAS distribution, v2004.

        Cleaned up batch.k shell script that runs McIDAS postprocess
        BATCH files.  Internal comments should now be easier to follow
        than in previous versions.

        Updated man pages for currently used decoders: area2png(1),
        nldn2md(1), pnga2area(1), pngg2gini(1), proftomd(1), and
        zlibg2gini(1).

        Removed building of decoders now considered to be obsolete:
        cdftomd(1), lwtoa3(1), gunrv2(1), lwfile(1), lwtmd2(1),
        and nids2area(1).  Source code for these decoders is still
        included in the distribution, but their use is not guaranteed
        mainly since subroutines they use from the McIDAS library
        are no longer supported.

2003    2003-07-29

        Added utililty routine zlibg2gini that uncompresses Zlib
        compressed GINI images in NIMAGE IDD datastream.

        Fixed a bug that kept McIDAS routing table, ROUTE.SYS,
        dates and times from being updated.

2002    2002-12-20

        Replaced information messages in pngg2gini.c that referred to
        pnga2area with pngg2gini.

        Corrected logic in pnga2area.c for processing a product whose
        product code is not contained in the McIDAS routing table,
        ROUTE.SYS.  Before, the routing table would get corrupted
        and the AREA file number included in the '-r pcode,anum'
        flag would not be used.  Now, the routing table will not
        be updated for unregistered products, and the default AREA
        file number will be used.
	

decode/lgt2md.c
/*
** Name:      lgt2md.c
**
** Purpose:   Lightning decoder for NLDN and USPLN data.
**
** Note:      See usage information for options.
**
** Copyright: Copyright 1997, University Corporation for Atmospheric Research
**            Do not redistribute without permission.
**
** History:   20060923 - created
*/

/*
** Includes
*/
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include "ulog.h"

#ifdef NO_ATEXIT
#  include <atexit.h>
#endif

#include "lightning.h"

/*
** Function prototypes
*/
int  decode_lgt( char *, int, char *, char *, int, int );

/*
** Usage
*/

static void usage(argv0)
char *argv0;
{
#define USAGE_FMT "\
Usage: %s [flags] pcode mdbase\n\
Flags:\n\
    -h           Print the help file, then exit the program\n\
    -f filename  Input NetCDF file (default is stdin)\n\
    -n filename  Output NetCDF file name (default not saved)\n\
                 [-n is ignored if -f is used]\n\
    -d path      Change to directory <path> before decoding\n\
    -l filename  Log output (default uses local0, '-' to stdout\n\
    -v           Verbose logging\n\
    -x           Debug logging\n\
    -e VAR=value Set an environmental variable in program\n\
Positionals:\n\
    mdbase       McIDAS base MD file number (91 - 6 min, 81 - hourly)\n\
    schema       McIDAS MD file schema (WPR6 - 6 min, WPRO - hourly)\n\
Keywords:\n\
    DIALPROD=pcode [day] [time]\n\
             pcode - McIDAS product code (LD - NLDN, LP - USPLN)\n\
             day   - day [ccyyjjj]\n\
             time  - time [hhmmss]\n\
    DEV=xyz\n\
             Ignored as of ldm-mcidas v2006\n"

    (void) fprintf (stderr, USAGE_FMT, argv0);
}

/*
** This callback routine registered by atexit().
*/
static void cleanup() {
    uinfo( "Exiting" ) ;
    (void) closeulog();
}


typedef struct envlist{
    char *env;
    struct envlist *next;
} envlist;


main(argc,argv)
int   argc;
char *argv[];
{

    char         *datadir   = NULL;
    char         *infilnam  = NULL;
    char         *mcpathbuf = NULL;       /* MCPATH environment variable ptr */
    char         *logfname  = NULL;
    char          defdir[]  = "./";

    /* Command line flag parsing */
    int           ch;
    extern int    optind;
    extern int    opterr;
    extern char  *optarg;
    envlist      *envhead=NULL;
    envlist      *envobj;

    /* Command line parameters */
    int           mdbase;
    char         *schema;
    char          pcode[3];               /* DIALPROD=pcode day time */
    int           rday;                   /* DIALPROD=pcode day time */
    int           rtime;                  /* DIALPROD=pcode day time */

    /* Logging */
    int           logmask = (LOG_MASK(LOG_ERR) | LOG_MASK(LOG_NOTICE));
    
    lgt_struct   *data=NULL;
    int           rc;
    
    /*
     * register exit handler
     */
    if ( atexit(cleanup) != 0 ) {
      serror( "atexit" );
      exit( 1 );
    }
    
    /*
    ** Parse command line flags
    */
    opterr = 1;
    while ((ch = getopt(argc,argv,"hvxl:d:f:n:e:")) != EOF) {
       switch(ch) {
          case 'h':
             usage(argv[0]);
             exit(0);
             break;
          case 'd':
             datadir  = optarg;
             break;
          case 'f':
             infilnam = optarg;
             break;
          case 'l':
             logfname = optarg;
             break;
          case 'v':
             logmask |= LOG_MASK(LOG_INFO);
             break;
          case 'x':
             logmask |= LOG_MASK(LOG_DEBUG);
             break;
          case 'e':
             envobj       = (envlist *) malloc( sizeof(envlist) );
             envobj->env  = malloc( strlen(optarg) );
             strcpy( envobj->env,optarg );
             envobj->next = envhead;
             envhead      = envobj;
             break;
          case '?':
             usage(argv[0]);
             exit(0);
          }
    }
    if ( datadir == (char *) NULL ) {
      datadir = defdir;
    }
    
    /*
    ** Parse command line positional parameters
    */
    switch( argc-optind ) {
       case 2:
       case 3:
       case 4:
       case 5:
       case 6:
          mdbase = atol(argv[optind++]);
          mdbase = 10 * (mdbase / 10);
          schema = argv[optind++];
          break;
       default:
          usage(argv[0]);
          exit(-1);
    }
    
    /*
    ** Parse command line keyword parameters
    */
    switch( argc-optind ) {
       case 0:
          (void) strcpy( pcode, "LD" );
          rday   = -1;
          rtime  = -1;
          break;
       case 1:
          (void) sscanf( argv[optind++], "DIALPROD=%2s", pcode );
          break;
       case 2:
          (void) sscanf( argv[optind++], "DIALPROD=%2s", pcode );
          rday  = atol( argv[optind++] );
          break;
       case 3:
       case 4:
          (void) sscanf( argv[optind++], "DIALPROD=%2s", pcode );
          rday  = atol( argv[optind++] );
          rtime = atol( argv[optind++] );
          break;
       default:
          usage(argv[0]);
          exit(-1);
    }

    /*
    ** Set ulog mask and announce start
    */
    (void) setulogmask(logmask);
    (void) openulog(ubasename(argv[0]),(LOG_CONS|LOG_PID),LOG_LOCAL0,logfname);
    uinfo("Starting up");
    
    udebug( "MD base = %4d\n", mdbase );
    udebug( "SCHEMA = %s\n", schema );
    udebug( "Pcode = %s\n", pcode );
    udebug( "Day = %d\n", rday );
    udebug( "Time = %d\n", rtime );

	/*
	 * Unset MCPATH environment variable if it exists
	 */
	mcpathbuf = getenv( "MCPATH" );
	if ( mcpathbuf != (char *)NULL ) {
	  uinfo( "unsetting MCPATH environment variable" );
	  (void) putenv( "MCPATH=" );
	}

    /*
    ** Set environment variable if '-e' flag set by user
    */
    envobj = envhead;
    while ( envobj != (envlist *)NULL ) {
      udebug( "putenv %s\0", envobj->env );
      if ( putenv(envobj->env) != 0 )
        uerror( "Can't set environmental variable %s\0", envobj->env );
      envobj = envobj->next;
    }
    
    /*
    ** If input is from a file redirect STDIN
    */
    if ( infilnam != (char *)NULL ) {
       if ( freopen(infilnam, "r", stdin) == NULL ) {
         uerror( "Unable to open input file \"%s\", exiting", infilnam );
         exit( 1 );
       }
       udebug( "input data from %s\0", infilnam );
    } else {
       udebug( "input data from STDIN" );
    }
    
    /*
    ** Call lightning decoder passing command line parameters
    */
    rc = decode_lgt( datadir, mdbase, schema, pcode, rday, rtime );
    
    /*
    ** Done
    */
    exit( 0 );

}
